sympy/physics/quantum/qapply.py

```diff
@@ -28,6 +29,17 @@
 # Main code
 #-----------------------------------------------------------------------------
 
+
+def ip_doit_func(e):
+    """Transform the inner products in an expression by calling ``.doit()``."""
+    return e.replace(InnerProduct, lambda *args: InnerProduct(*args).doit())
+
+
+def sum_doit_func(e):
+    """Transform the sums in an expression by calling ``.doit()``."""
+    return e.replace(Sum, lambda *args: Sum(*args).doit())
```

Comment:
There should really be a general way of doing this. Pretty much all library code that calls `.doit()` really should only call `.doit()` on some particular type like this.